test/cmp_hdr_test.c

```diff
@@ -0,0 +1,472 @@
+    return 1;
+}
+
+static int test_HDR_get0_senderNonce(void)
+{
+    SETUP_TEST_FIXTURE(CMP_HDR_TEST_FIXTURE, set_up);
+    fixture->expected = 1;
+    EXECUTE_TEST(execute_HDR_get0_senderNonce_test, tear_down);
+    return result;
+}
+
+static int execute_HDR_set1_sender_test(CMP_HDR_TEST_FIXTURE *fixture)
+{
+    X509_NAME *x509name = X509_NAME_new();
+
+    X509_NAME_add_entry_by_txt(x509name, "CN",
+          MBSTRING_ASC, (unsigned char*)"A common sender name", -1, -1, 0);
+    if (!TEST_int_eq(ossl_cmp_hdr_set1_sender(fixture->hdr, x509name), 1)) {
+        return 0;
+    };
+    if (!TEST_int_eq(fixture->hdr->sender->type, GEN_DIRNAME)) {
+        return 0;
+    }
+    if (!TEST_int_eq(
+            X509_NAME_cmp(fixture->hdr->sender->d.directoryName, x509name),
+                          0)) {
+        return 0;
+    }
+    X509_NAME_free(x509name);
+    return 1;
+}
+
+static int test_HDR_set1_sender(void)
+{
+    SETUP_TEST_FIXTURE(CMP_HDR_TEST_FIXTURE, set_up);
+    fixture->expected = 1;
+    EXECUTE_TEST(execute_HDR_set1_sender_test, tear_down);
+    return result;
+}
+
+static int execute_HDR_set1_recipient_test(CMP_HDR_TEST_FIXTURE *fixture)
+{
+    X509_NAME *x509name = X509_NAME_new();
+
+    X509_NAME_add_entry_by_txt(x509name, "CN",
+          MBSTRING_ASC, (unsigned char*)"A common recipient name", -1, -1, 0);
+    if (!TEST_int_eq(ossl_cmp_hdr_set1_recipient(fixture->hdr, x509name), 1)) {
+        return 0;
+    };
+    if (!TEST_int_eq(fixture->hdr->recipient->type, GEN_DIRNAME)) {
+        return 0;
+    }
+    if (!TEST_int_eq(
+            X509_NAME_cmp(fixture->hdr->recipient->d.directoryName, x509name),
+                          0)) {
+        return 0;
+    }
+    X509_NAME_free(x509name);
+    return 1;
+}
+
+static int test_HDR_set1_recipient(void)
+{
+    SETUP_TEST_FIXTURE(CMP_HDR_TEST_FIXTURE, set_up);
+    fixture->expected = 1;
+    EXECUTE_TEST(execute_HDR_set1_recipient_test, tear_down);
+    return result;
+}
+
+static int execute_HDR_update_messageTime_test(CMP_HDR_TEST_FIXTURE *fixture)
+{
+    struct tm tm;
+    time_t t1, t2, now;
+    double diffSecs;
+
+    if (!TEST_int_eq(ossl_cmp_hdr_update_messageTime(fixture->hdr), 1)) {
+        return 0;
+    };
+    if (!TEST_int_eq(ASN1_TIME_to_tm(fixture->hdr->messageTime, &tm), 1)) {
+        return 0;
+    };
+    t1 = mktime(&tm);
+    now = time(NULL);
+    t2 = mktime(gmtime(&now));
+    diffSecs = difftime(t1, t2);
+    if (!TEST_true(diffSecs > -2.0 && diffSecs < +2.0)) {
+        return 0;
+    }
+
+    return 1;
+}
+
+static int test_HDR_update_messageTime(void)
+{
+    SETUP_TEST_FIXTURE(CMP_HDR_TEST_FIXTURE, set_up);
+    fixture->expected = 1;
+    EXECUTE_TEST(execute_HDR_update_messageTime_test, tear_down);
+    return result;
+}
+
+static int execute_HDR_set1_senderKID_test(CMP_HDR_TEST_FIXTURE *fixture)
+{
+    ASN1_OCTET_STRING* senderKID = ASN1_OCTET_STRING_new();
+
+    ASN1_OCTET_STRING_set(senderKID, rand_data, sizeof(rand_data));
```

Comment:
rand_data never seems to be initialised to anything, so I guess its just all zeros. Would it better to actually have non-zero data in it?

Reply:
We now call RAND_bytes() before